Implements a fix for ADAL cache replication

iambmelt opened the pull request:
Testing still in progress, but the outline of a fix for the below bug:
- https://identitydivision.visualstudio.com/Engineering/_workitems/edit/1437575

iambmelt commented on adal/src/main/java/com/microsoft/aad/adal/DelegatingCache.java:
```diff
@@ -0,0 +1,81 @@
+
+import java.util.Iterator;
+
+class DelegatingCache implements ITokenCacheStore {
```
TODO javadoc

iambmelt commented on adal/src/main/java/com/microsoft/aad/adal/TokenCacheAccessor.java:
```diff
@@ -98,10 +98,11 @@
             //If the default token cache is in use... delegate token operations to unified cache in common
             //If not using default token cache then sharing SSO state between ADAL & MSAL cache implementations will not be possible anyway
-            mUseCommonCache = true;
+            mUseCommonCache = delegate instanceof DefaultTokenCacheStore;
```
Consider using `isAssignableFrom`

AdamBJohnsonx replied:
Is it possible to write a unit test here that highlights the problem, and then is fixed by this code change?

iambmelt replied:
21b0ce12526b95bb6a803b75b059282e31c6272e

iambmelt commented on adal/src/main/java/com/microsoft/aad/adal/DelegatingCache.java:
```diff
@@ -0,0 +1,81 @@
+    private final Context mContext;
+    private final ITokenCacheStore mDelegate;
+
+    DelegatingCache(@NonNull final Context context,
```
Javadoc

iambmelt commented on adal/src/main/java/com/microsoft/aad/adal/DelegatingCache.java:
```diff
@@ -0,0 +1,81 @@
+        mDelegate = delegate;
+    }
+
+    public ITokenCacheStore getDelegateCache() {
```
Javadoc

iambmelt commented:
> This test is detecting a particular kind of problem. Putting a javadoc block explaining what it was helps people reading the test without naming it something like testMsalCacheIsUpdatedWhenAdalCacheIsPresentAndAWriteGoesToTheLatter...
> 
> I think that's what we're doing...

Added javadoc, constants, cleaned up tests

AdamBJohnsonx commented on adal/src/androidTest/java/com/microsoft/aad/adal/TokenCacheAccessorTests.java:
```diff
@@ -169,14 +209,42 @@ public void setUp() throws Exception {
         );
+
+
+        // Populate a mock Instance Discovery
```
```suggestion
        // Populate a mock Instance Discovery containing entries for the common authority as
        // well as one instance for Mooncake.
```

iambmelt commented:
One of the ADAL tests is failing... looking into it to see if related to this PR

AdamBJohnsonx commented on adal/src/androidTest/java/com/microsoft/aad/adal/TokenCacheAccessorTests.java:
```diff
@@ -204,58 +271,130 @@ public void testUpdateTokenCacheUsesResultAuthority() throws MalformedURLExcepti
+        result.setAuthority(WORLDWIDE_AUTHORITY);
+        result.setClientInfo(new ClientInfo(MOCK_CLIENT_INFO));
+        result.setResponseReceived(System.currentTimeMillis());
+        result.setExpiresIn(System.currentTimeMillis());
```
That's a weird way to set expiresIn..

iambmelt commented:
> One of the ADAL tests is failing... looking into it to see if related to this PR

Root caused the issue to this constant
https://github.com/AzureAD/microsoft-authentication-library-common-for-android/blame/dev/common/src/main/java/com/microsoft/identity/common/internal/cache/ADALOAuth2TokenCache.java#L122

iambmelt commented:
> > One of the ADAL tests is failing... looking into it to see if related to this PR
> 
> Root caused the issue to this line
> https://github.com/AzureAD/microsoft-authentication-library-common-for-android/blame/dev/common/src/main/java/com/microsoft/identity/common/internal/cache/ADALOAuth2TokenCache.java#L122

Opened up a PR to resolve the issue here:
https://github.com/AzureAD/microsoft-authentication-library-common-for-android/pull/1444

AdamBJohnsonx reviewed (approved):
Only comment, is it expiresIn or expiresAt?  Setting expiresIn to the currentTime in milliseconds seems… strange.
